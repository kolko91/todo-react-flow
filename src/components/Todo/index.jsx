import React, { PureComponent } from 'react';
import styled from 'styled-components';

import TodoEditTextInput from 'components/TodoEditTextInput';
import type { Text } from 'types/todos';
import toogleImage from './toogle.svg';
import toogleCheckImage from './toogle-check.svg';

export type Props = {
  onClick: () => void,
  onDeleteClick: ()=> void,
  editTodo: (text: Text)=>void,
  completed: boolean,
  text: Text
};

const Toggle = styled.input`
    text-align: center;
    height: 40px;
    width: 40px;
    position: absolute;
    top: 0;
    bottom: 0;
    margin: auto 0;
    border: none;
    appearance: none;
    background: none;
    &:after {
    content: url(${toogleImage});
    }
    &:checked:after {
    content: url(${toogleCheckImage});
    }
    &:focus{
        outline:0;
    }
`;

const Label = styled.label`
    white-space: pre-line;
    word-break: break-all;
    padding: 15px 60px 15px 15px;
    margin-left: 45px;
    display: block;
    line-height: 1.2;
    transition: color 0.4s;
`;

const DestroyButton = styled.button`
    display: none;
    position: absolute;
    top: 0;
    right: 10px;
    bottom: 0;
    width: 40px;
    height: 40px;
    margin: auto 0;
    font-size: 30px;
    color: #cc9a9a;
    margin-bottom: 11px;
    transition: color 0.2s ease-out;
    &:after{
      content: 'Ã—';
    }
`;

const Item = styled.li`
    background: ${({ completed }: Props) => (completed ? '#f1f1f1' : '#fff')};
    text-decoration: ${({ completed }: Props) => (completed ? 'line-through' : 'none')};
    position: relative;
    font-size: 24px;
    border-bottom: 1px solid #ededed;
    &:hover ${DestroyButton}{
      display: block;
      color: #af5b5e;
    }
`;


class Todo extends PureComponent<Props, {editing: boolean}> {
  state = {
    editing: false,
  }

  handleDoubleClick = () => {
    this.setState({ editing: true });
  }

  handleSave = (text: Text) => {
    if (text.length === 0) {
      this.props.onDeleteClick();
    } else {
      this.props.editTodo(text);
    }
    this.setState({ editing: false });
  }

  render() {
    const {
      onClick, completed, text, onDeleteClick,
    } = this.props;
    const { editing } = this.state;
    let element;

    if (editing) {
      element = (
        <TodoEditTextInput
          text={text}
          onSave={(t: Text) => this.handleSave(t)}
        />
      );
    } else {
      element = (
        <div>
          <Toggle
            checked={completed}
            onChange={onClick}
            type="checkbox"
          />
          <Label onDoubleClick={this.handleDoubleClick}>{text}</Label>
          <DestroyButton onClick={onDeleteClick} />
        </div>);
    }

    return (
      <Item completed={completed} >
        {element}
      </Item>);
  }
}

export default Todo;
