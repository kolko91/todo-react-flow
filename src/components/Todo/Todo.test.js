import React from 'react';
import renderer from 'react-test-renderer';
import Todo from './index';


describe('component', () => {
  describe('Todo', () => {
    it('should render correctly', () => {
      const props = {
        onClick: jest.fn(),
        onDeleteClick: jest.fn(),
        editTodo: jest.fn(),
        done: false,
        title: 'Test',
      };
      const tree = renderer.create(<Todo {...props} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
