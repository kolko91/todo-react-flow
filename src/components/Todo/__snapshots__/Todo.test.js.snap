// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component Todo should render correctly 1`] = `
<li
  className="sc-ifAKCX kYTDwW"
>
  <div>
    <input
      checked={false}
      className="sc-bwzfXH dydaOf"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="sc-htpNat dDFmjJ"
      onDoubleClick={[Function]}
    >
      Test
    </label>
    <button
      className="sc-bxivhb bKUsik"
      onClick={[Function]}
    />
  </div>
</li>
`;
